x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportUnfollowAction.java

```diff
@@ -0,0 +1,116 @@
+import org.elasticsearch.xpack.ccr.CcrSettings;
+import org.elasticsearch.xpack.core.ccr.action.UnfollowAction;
+
+public class TransportUnfollowAction extends TransportMasterNodeAction<UnfollowAction.Request, AcknowledgedResponse> {
+
+    @Inject
+    public TransportUnfollowAction(Settings settings, TransportService transportService, ClusterService clusterService,
+                                   ThreadPool threadPool, ActionFilters actionFilters,
+                                   IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(settings, UnfollowAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            UnfollowAction.Request::new, indexNameExpressionResolver);
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected AcknowledgedResponse newResponse() {
+        return new AcknowledgedResponse();
+    }
+
+    @Override
+    protected void masterOperation(UnfollowAction.Request request,
+                                   ClusterState state,
+                                   ActionListener<AcknowledgedResponse> listener) throws Exception {
+        clusterService.submitStateUpdateTask("unfollow_action", new ClusterStateUpdateTask() {
+
+            @Override
+            public ClusterState execute(ClusterState current) throws Exception {
+                String followerIndex = request.getFollowerIndex();
+                return unfollow(followerIndex, current);
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                listener.onResponse(new AcknowledgedResponse(true));
+            }
+        });
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(UnfollowAction.Request request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+    }
+
+    static ClusterState unfollow(String followerIndex, ClusterState current) {
+        IndexMetaData followerIMD = current.metaData().index(followerIndex);
+
+        PersistentTasksCustomMetaData persistentTasks = current.metaData().custom(PersistentTasksCustomMetaData.TYPE);
+        if (persistentTasks != null) {
+            for (PersistentTasksCustomMetaData.PersistentTask<?> persistentTask : persistentTasks.tasks()) {
+                if (persistentTask.getTaskName().equals(ShardFollowTask.NAME)) {
+                    ShardFollowTask shardFollowTask = (ShardFollowTask) persistentTask.getParams();
+                    if (shardFollowTask.getFollowShardId().getIndexName().equals(followerIndex)) {
+                        throw new IllegalArgumentException("cannot unfollow index [" + followerIndex +
```

Comment:
I think saying that we can not convert the follower index to a non-follower would be clearer. My concern here is that if `bar` is following `foo` and this message says `cannot unfollow index [bar]` it would be confusing since it is `foo` that will no longer be being followed by `bar`.